### PR TITLE
kubelet/config: fail pod decode if hostUsers=false but UserNamespacesSupport is off

### LINKED ISSUE #126484: kubelet would happily schedule a pod with `spec.hostUsers: false` even if `UserNamespacesSupport` is not set for kubelet
### What happened?

As per https://github.com/kubernetes/kubernetes/pull/123216/files#diff-ed935ef85e4cd93302c07762fbc4314164759fb618bd661a3ac17ca782dde836R397 I would have expected kubelet to disallow a pod that has `spec.hostUsers: false` set when the `UserNamespacesSupport` is not enabled for the kubelet.

From testing kubelet and api server would happily accept a pod having the spec and gives the user a false sense of security.

From my further testing if the `UserNamespacesSupport` feature gate is set for api server and not set for kubelet the pod fails to be scheduled.

### What did you expect to happen?

Expected to see the error message below:

```text
Failed to create pod sandbox: the feature gate "UserNamespacesSupport" is disabled: can't set spec.HostUsers
```

when `UserNamespacesSupport` is not enabled at all for any component of kubernetes

### How can we reproduce it (as minimally and precisely as possible)?

Schedule a pod on a standard k8s cluster that has the `UserNamespacesSupport` feature gate not set for both kubelet and kube-apiserver

Sample pod spec used:

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: nginx
  name: nginx
spec:
  hostUsers: false
  containers:
  - image: nginx
    name: nginx
    resources: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Always
status: {}

```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
❯ kubectl version
Client Version: v1.30.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.31.0-beta.0
```

</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Talos"
ID=talos
VERSION_ID=v1.8.0-alpha.1-40-g96492c097
PRETTY_NAME="Talos (v1.8.0-alpha.1-40-g96492c097)"
HOME_URL="https://www.talos.dev/"
BUG_REPORT_URL="https://github.com/siderolabs/talos/issues"
VENDOR_NAME="Sidero Labs"
VENDOR_URL="https://www.siderolabs.com/"
$ uname -a
Linux version 6.6.43-talos (@buildkitsandbox) (gcc (GCC) 13.3.0, GNU ld (GNU Binutils) 2.42) #1 SMP Mon Jul 29 17:04:08 UTC 2024
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>
```
# runc --version
runc version v1.2.0-rc.2
commit: f2d2ee5e45af3c9b38dfd639ac92f43c3055ce4e
spec: 1.2.0
go: go1.22.5
libseccomp: 2.5.5
```
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
This is to protect against situations where the kube-apiserver relaxes validation for a pod assuming it is confined by a user namespace when it is not.

Do so here instead of in pkg/apis to avoid leaking this behavior to the kube-apiserver, which should continue to filter the field if the feature gate is off
#### Which issue(s) this PR fixes:

fixes https://github.com/kubernetes/kubernetes/issues/126484
#### Does this PR introduce a user-facing change?

```release-note
kubelet: fix a bug where the kubelet allows a pod with `hostUsers: false` to be created when it doesn't have the `UserNamespacesSupport` feature on. Previously the pod would be created without a pod level user namespace, which would break assumptions the kube-apiserver makes in relaxing the validation for the pod because it assumes the user namespace was created. Now, the kubelet fails to decode this pod, and the pod isn't created.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by haircommander** [comment]
cc @liggitt  @deads2k

File: pkg/kubelet/config/common.go [hunk 1]
```diff
@@ -132,6 +138,9 @@ func tryDecodeSinglePod(data []byte, defaultFn defaultFunc) (parsed bool, pod *v
 	if err = defaultFn(newPod); err != nil {
 		return true, pod, err
 	}
+	if err = validatePodFeatures(newPod); err != nil {
```
Comment (liggitt) [inline pkg/kubelet/config/common.go:141]: this is only used for static pods / pods from a url manifest, not pods coming from the API server ... I don't think this is where you want to add this

**Comment by liggitt** [comment]
I don't think this does what you think it does ... the guard should definitely only be in the kubelet, but not in tryDecodeSinglePod which only applies to static pods

Including the feature check in GetOrCreateUserNamespaceMappings seems more correct to me ... can you describe why that wasn't working as expected?

**Comment by haircommander** [comment]
/hold

we're already [failing there](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/userns/userns_manager.go#L402) but I haven't seen the failure, which implies to me we're calling ValidatePodCreate or some similar function that filters out fields that rely on the feature... it's possible I've not tested the right feature gate configuration though I need to check that

**Comment by haircommander** [comment]
```
Failed to create pod sandbox: the feature gate "UserNamespacesSupport" is disabled: can't set spec.HostUsers
```
silly me, I'm not sure what I messed up, but we don't need this. also, looking at https://github.com/kubernetes/kubernetes/issues/126484, the reporter has a different issue

/close

again :upside_down_face:

**Pull request closed without merging**